benchmark/paddle/CARLA_SAC/train.py

```diff
@@ -0,0 +1,153 @@
+    train_envs_params = EnvConfig['train_envs_params']
+    env_list = ParallelEnv(args.env, args.xparl_addr, train_envs_params)
+
+    # env for eval
+    eval_env_params = EnvConfig['eval_env_params']
+    eval_env = LocalEnv(args.env, eval_env_params)
+
+    obs_dim = eval_env.obs_dim
+    action_dim = eval_env.action_dim
+
+    # Initialize model, algorithm, agent, replay_memory
+    model = CarlaModel(obs_dim, action_dim)
+    algorithm = SAC(
+        model,
+        gamma=GAMMA,
+        tau=TAU,
+        alpha=ALPHA,
+        actor_lr=ACTOR_LR,
+        critic_lr=CRITIC_LR)
+    agent = CarlaAgent(algorithm)
+    rpm = ReplayMemory(
+        max_size=MEMORY_SIZE, obs_dim=obs_dim, act_dim=action_dim)
+
+    total_steps = 0
+    last_save_steps = 0
+    test_flag = 0
+
+    obs_list = env_list.reset()
+
+    while total_steps < args.train_total_steps:
+        # Train episode
+        if rpm.size() < WARMUP_STEPS:
+            action_list = [
+                np.random.uniform(-1, 1, size=action_dim)
+                for _ in range(len(train_envs_params))
+            ]
+        else:
+            action_list = [agent.sample(obs) for obs in obs_list]
+        next_obs_list, reward_list, done_list, info_list = env_list.step(
+            action_list)
+
+        # Store data in replay memory
+        for i in range(len(train_envs_params)):
+            rpm.append(obs_list[i], action_list[i], reward_list[i],
+                       next_obs_list[i], done_list[i])
+
+        obs_list = env_list.get_obs()
+        total_steps = env_list.total_steps
+        # Train agent after collecting sufficient data
+        if rpm.size() >= WARMUP_STEPS:
+            batch_obs, batch_action, batch_reward, batch_next_obs, batch_terminal = rpm.sample_batch(
+                BATCH_SIZE)
+            agent.learn(batch_obs, batch_action, batch_reward, batch_next_obs,
+                        batch_terminal)
+
+        # Save agent
+        if total_steps > int(1e5) and total_steps > last_save_steps + int(1e4):
+            agent.save('./model/step_{}_model.ckpt'.format(total_steps))
+            last_save_steps = total_steps
+
+        # Evaluate episode
+        if (total_steps + 1) // args.test_every_steps >= test_flag:
+            while (total_steps + 1) // args.test_every_steps >= test_flag:
+                test_flag += 1
+            avg_reward = run_evaluate_episodes(agent, eval_env, EVAL_EPISODES)
+            tensorboard.add_scalar('eval/episode_reward', avg_reward,
+                                   total_steps)
+            logger.info(
+                'Total steps {}, Evaluation over {} episodes, Average reward: {}'
+                .format(total_steps, EVAL_EPISODES, avg_reward))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--xparl_addr",
+        default='localhost:8080',
+        help='xparl address for parallel training')
+    parser.add_argument("--env", default="carla-v0")
```

Comment:
Is there any other environment in carla? Why do we put this argument here?

Reply:
The default environment uses a reconstructed list of a RGB camera's view as observation, while users can use other sensors to get a more meaningful observation. In case users use a different environment.